axlearn/common/attention.py

```diff
@@ -843,44 +1042,32 @@ def extend_step(
         query_positions += time_step[:, None]
 
         # Project inputs to key, value and query. Each has shape [B, steps, N, H].
-        q_proj, k_proj, v_proj = self.forward(query, **kv_kwargs, query_positions=query_positions)
+        q_proj, k_proj, v_proj, unused_q_pos, unused_k_pos = self.forward(
+            query, query_positions=query_positions, **kv_kwargs
+        )
         updated_state = dict(time_step=time_step + num_query_steps)
-
         if kv_state is None:
-            # Update the cache via one-hot broadcast and addition.
-            # NB: Cache updates can also be done via dynamic slice update. However it was observed
-            # that RLHF training got stuck in some cases.
-            # TODO(ds-hwang): Investigate the root cause.
-            cached_key = cached_states["key"]
-            cached_value = cached_states["value"]
-
-            source_len = cached_key.shape[-1]
-
-            # [B, T, N, H] --> [B, N, H, T].
-            k_proj = jnp.einsum("btnh->bnht", k_proj)
-            v_proj = jnp.einsum("btnh->bnht", v_proj)
-
-            # Create a dispatch matrix of shape [B, T=step, S].
-            oh_indices = jax.nn.one_hot(
-                time_step[:, None] + jnp.arange(num_query_steps), source_len, dtype=cached_key.dtype
+            with child_context("kv_cache_extend_step", module=self.kv_cache):
```

Comment:
Do we need this child_context?

Reply:
Yes, unfortunately. It's because `init_states()` calls here while `init_states()` already called `kv_cache.init_states()`